metals/src/main/scala/scala/meta/internal/rename/RenameProvider.scala

```diff
@@ -67,173 +67,182 @@ final class RenameProvider(
+          .flatMap { _ =>
+            definitionProvider.definition(source, params, token).map {
+              definition =>
+                val symbolOccurrence: Option[(SymbolOccurrence, TextDocument)] =
+                  definitionProvider
+                    .symbolOccurrence(source, params.getPosition)
+                    .orElse(
+                      findRenamedImportOccurrenceAtPosition(
+                        source,
+                        params.getPosition(),
+                      )
+                    )
+                for {
+                  (occurence, _) <- symbolOccurrence
+                  soughtSymbols = Set(occurence.symbol) ++ companion(
+                    occurence.symbol
+                  )
+                  definitionLocation <- definition.locations.asScala.headOption
+                  definitionPath = definitionLocation.getUri().toAbsolutePath
+                  if canRenameSymbol(occurence.symbol, None) &&
+                    (isWorkspaceSymbol(occurence.symbol, definitionPath) ||
+                      findRenamedImportForSymbol(
+                        source,
+                        soughtSymbols,
+                        occurence.symbol.desc.name.value,
+                      ).isDefined)
+                  range <- occurence.range
+                } yield range.toLsp
+            }
+          }
+      )
   }
 
   def rename(
       params: RenameParams,
       token: CancelToken,
   ): Future[WorkspaceEdit] = {
     val source = params.getTextDocument.getUri.toAbsolutePath
-    compilations.compilationFinished(source).flatMap { _ =>
-      val defininionFuture = definitionProvider
-        .definition(source, params, token)
-      defininionFuture.flatMap { definition =>
-        // We have to check if local symbol doesn't override any non-local symbols
-        // If it does, we can't use PC for local rename
-        def overridesNonLocal = (for {
-          semantic <- definition.semanticdb
-          symbolInfo <- semantic.symbols
-            .find(_.symbol == definition.symbol)
-          overridenSymbols = symbolInfo.overriddenSymbols
-        } yield overridenSymbols.exists(!_.isLocal)).getOrElse(true)
-        if (definition.symbol.isLocal && !overridesNonLocal) {
-          compilers
-            .rename(params, token)
-            .map(_.asScala.toList)
-            .map { edits =>
-              new WorkspaceEdit(
-                documentEdits(Map(source -> edits)).asJava
-              )
-            }
-        } else {
-          val textParams = new TextDocumentPositionParams(
-            params.getTextDocument(),
-            params.getPosition(),
-          )
-
-          lazy val definitionTextParams =
-            definition.locations.asScala.map { l =>
-              new TextDocumentPositionParams(
-                new TextDocumentIdentifier(l.getUri()),
-                l.getRange().getStart(),
-              )
-            }
+    val localRename = compilers
```

Comment:
It looks like a lot of changes, but it's mostly because of new indentation. Important thing that changed, is that now we don't check if we can use presentation compiler here, but in `PcRenameProvider`